caffe2/opt/backend_cutting.cc

```diff
@@ -0,0 +1,390 @@
+#include "caffe2/opt/backend_cutting.h"
+#include "caffe2/core/logging.h"
+#include "nomnigraph/Converters/Caffe2.h"
+#include "nomnigraph/Converters/Dot.h"
+#include "nomnigraph/Representations/NeuralNet.h"
+
+#include <algorithm>
+#include <fstream>
+#include <queue>
+
+namespace caffe2 {
+namespace opt {
+
+namespace {
+
+using namespace nom::repr;
+using NodeRef = NNGraph::NodeRef;
+using EdgeRef = NNGraph::EdgeRef;
+
+class GroupAnnotation {
+ public:
+  GroupAnnotation(int i, int g = -1) : group(g), in_degree(i) {}
+  int group;
+  int in_degree;
+  bool needs_transform{true};
+};
+
+struct VisitorContext {
+  VisitorContext(std::function<bool(const caffe2::OperatorDef&)> func)
+      : predicate(func) {}
+  std::unordered_map<NodeRef, GroupAnnotation> infos;
+  std::unordered_set<NodeRef> frontier;
+  std::vector<NodeRef> current_group;
+  std::function<bool(const caffe2::OperatorDef&)> predicate;
+
+  int group{0};
+  bool find_supported{true};
+};
+
+std::string ShowNode(NodeRef node) {
+  if (nn::is<NeuralNetData>(node)) {
+    const auto* nn_tensor = nn::get<NeuralNetData>(node); 
+    return MakeString("Tensor: ", nn_tensor->getName());
+  } else if (nn::is<NeuralNetOperator>(node)) {
+    const auto* nn_op = nn::get<NeuralNetOperator>(node); 
+    const auto* op_def = reinterpret_cast<const caffe2::OperatorDef*>(
+        nn_op->getAnnotation()->getSaved());
+    CAFFE_ENFORCE(op_def);
+    return MakeString("Op: ", op_def->type());
+  } else {
+    CAFFE_THROW("Known node");
+  }
+}
+
+void DumpGraph(NNGraph* g) {
+  auto nnprinter = [](typename NNGraph::NodeRef node) {
+    std::map<std::string, std::string> labelMap;
+    assert(node->data() && "Node doesn't have data, can't render it");
+    if (isa<NeuralNetOperator>(node->data())) {
+      auto* op = dyn_cast<NeuralNetOperator>(node->data().get());
+      labelMap["label"] =
+          op->getName() + " (" + std::to_string((unsigned long long)node) + ")";
```

Comment:
you'll need to use caffe2' to_string (there is an issue with to_string on android that is breaking the build: https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-android-ubuntu16.04-build/2863/console)